RecoLocalCalo/EcalRecAlgos/src/KernelHelpers.cu

```diff
@@ -4,49 +4,215 @@
 #include "DataFormats/EcalDetId/interface/EEDetId.h"
 
 namespace ecal {
-  namespace multifit {
-
+  namespace reconstruction {
+    
     namespace internal {
-
+      
       namespace barrel {
-
+        
         __device__ __forceinline__ bool positiveZ(uint32_t id) { return id & 0x10000; }
-
+        
         __device__ __forceinline__ uint32_t ietaAbs(uint32_t id) { return (id >> 9) & 0x7F; }
-
+        
         __device__ __forceinline__ uint32_t iphi(uint32_t id) { return id & 0x1FF; }
-
+        
+        
+        
+        __device__ int dccFromSm(int ism) {
+          int iz = 1;
+          if (ism > 18)
+            iz = -1;
+          if (iz == -1)
+            ism -= 18;
+          int idcc = 9 + ism;
+          if (iz == +1)
+            idcc += 18;
+          return idcc;  
+        }
+        
+        __device__ int sm(int ieta, int iphi) {
+          int iz = 1;
+          if (ieta < 0)
+            iz = -1;
+          ieta *= iz;
+          int iphi_ = iphi;
+          if (iphi_ > 360)
+            iphi_ -= 360;
+          int ism = (iphi_ - 1) / 20 + 1;
+          if (iz == -1)
+            ism += 18;
+          return ism;
+        }
+        
+        
+        __device__ int dcc(int ieta, int iphi) {
+          int ism = sm(ieta, iphi);
+          return dccFromSm(ism);
+        }
+        
+        
+        
+        
+        //        
+        // ---- why on hell things are so complex and not simple ???
+        //        
+        
+        
+        __device__ int lm_channel (int iX, int iY) {
+          
+          static const int idx_[] = {
+            // 0  1  2  3  4  5  6  7  8  9 10 11 12 13 14 15 16
+            1, 2, 2, 2, 2, 4, 4, 4, 4,
+            6, 6, 6, 6, 8, 8, 8, 8,  // 3
+            1, 2, 2, 2, 2, 4, 4, 4, 4,
+            6, 6, 6, 6, 8, 8, 8, 8,  // 2
+            1, 3, 3, 3, 3, 5, 5, 5, 5,
+            7, 7, 7, 7, 9, 9, 9, 9,  // 1
+            1, 3, 3, 3, 3, 5, 5, 5, 5,
+            7, 7, 7, 7, 9, 9, 9, 9  // 0
+            // 0  1  2  3  4  5  6  7  8  9 10 11 12 13 14 15 16
+          };
+          
+          int iym, ixm, il, ic, ii;
+          iym = 4;
+          ixm = 17;
+          int iX_ = iX + 1;
+          int iY_ = iY + 1;
+          il = iym - iY_;
+          ic = iX_ - 1;
```

Comment:
isn't this the same as `iX` ?

Reply:
honestly yes ... 
but as before I would suggest to have a separate PR with all the changes

https://github.com/cms-sw/cmssw/blob/2c91040e71a5447c0e161e80ce914260713b8317/CalibCalorimetry/EcalLaserAnalyzer/src/MEEBGeom.cc#L215-L226

gpu and cpu, changed consistently.